ADApp/ADSrc/NDAttribute.cpp

```diff
@@ -388,4 +388,25 @@ int NDAttribute::report(FILE *fp, int details)
   return ND_SUCCESS;
 }
 
+/** Get the max number of bits per pixel for the specified data type
+  * \param[in] NDDataType_t  Data type
+  */
+int GetNDDataTypeBits( NDDataType_t tyData )
+{
+    int nBits;
+    switch ( tyData )
+    {
+    // Default to large number so conversions to
+    // integer data types will be clipped to max
+    default:        nBits   = 99;   break;
+    case NDInt8:    nBits   = 8;    break;
+    case NDUInt8:   nBits   = 8;    break;
+    case NDInt16:   nBits   = 16;   break;
+    case NDUInt16:  nBits   = 16;   break;
+    case NDInt32:   nBits   = 32;   break;
+    case NDUInt32:  nBits   = 32;   break;
+    }
```

Comment:
> I'm not aware of any color image standards that use anything but 8 or 16 except for digital video where you may see more bits for

> luminance, for example, 6 bits G, 5 bits R and 5 bits B. For color, the bitsPerPixel parameter is of less interest as the color mode tells you all you need to know.

There may well be cameras that produce 16-bit RGB, but for which it only uses the low-order 12 bits for example.

There are also the YUV color modes that are supported in NDArray.h.  They can encode 2 RGB pixels in 4 bytes for example, or 4 pixels in 6 bytes.
typedef enum
{
    NDColorModeMono,    /**< Monochromatic image */
    NDColorModeBayer,   /**< Bayer pattern image, 1 value per pixel but with color filter on detector */
    NDColorModeRGB1,    /**< RGB image with pixel color interleave, data array is [3, NX, NY] */
    NDColorModeRGB2,    /**< RGB image with row color interleave, data array is [NX, 3, NY]  */
    NDColorModeRGB3,    /**< RGB image with plane color interleave, data array is [NX, NY, 3]  */
    NDColorModeYUV444,  /**< YUV image, 3 bytes encodes 1 RGB pixel */
    NDColorModeYUV422,  /**< YUV image, 4 bytes encodes 2 RGB pixel */
    NDColorModeYUV411   /**< YUV image, 6 bytes encodes 4 RGB pixels */
} NDColorMode_t;

Mark


From: bhill-slac [mailto:notifications@github.com]
Sent: Friday, January 16, 2015 4:19 PM
To: areaDetector/ADCore
Cc: Mark Rivers
Subject: Re: [ADCore] Added support for BitsPerPixel and BytesPerPixel in ADBase (#53)


In ADApp/ADSrc/NDAttribute.cpp<https://github.com/areaDetector/ADCore/pull/53#discussion-diff-23113653>:

> +  */

> +int GetNDDataTypeBits( NDDataType_t tyData )

> +{

> +    int nBits;

> +    switch ( tyData )

> +    {

> +    // Default to large number so conversions to

> +    // integer data types will be clipped to max

> +    default:        nBits   = 99;   break;

> +    case NDInt8:    nBits   = 8;    break;

> +    case NDUInt8:   nBits   = 8;    break;

> +    case NDInt16:   nBits   = 16;   break;

> +    case NDUInt16:  nBits   = 16;   break;

> +    case NDInt32:   nBits   = 32;   break;

> +    case NDUInt32:  nBits   = 32;   break;

> +    }

GetNDDataTypeBits() is used to determine the default value for newly alloced NDArray's and for clipping in NDArrayPool::convert when converting to a smaller data type. I agree it would be good to return -1 for invalid and specific values for NDFloat32 and NDFloat64. I'm also open to returning a different value than 99 as long as it's large. I don't think bitsPerPixel is particularly relevant for floating point NDArrays.

For color images, my searches seemed to indicate that discussions re the number of bits refer to the number of bits per pixel color, so 8 for typical RGB and 16 for RGB16. I'm not aware of any color image standards that use anything but 8 or 16 except for digital video where you may see more bits for luminance, for example, 6 bits G, 5 bits R and 5 bits B. For color, the
bitsPerPixel parameter is of less interest as the color mode tells you all you need to know.

Where bitsPerPixel is needed is for mono images, where there is no standard. We have mono cameras with 8, 10, 12, and 16 bits and sometimes use the colorConvert plugin to reduce waveform size by converting to Uint8 images.

To be useful, BitsPerPixel must be set by the driver. I've done this in one of my ADProsilica pull requests and in the ADEdtPdv driver I'll soon be posting for EDT framegrabber based cameras.

—
Reply to this email directly or view it on GitHub<https://github.com/areaDetector/ADCore/pull/53/files#r23113653>.